### PR TITLE
MRG: Unify _prepare_gain

### PR DESCRIPTION
**Pull request by larsoner**
Okay @agramfort I had to change a tolerance for one gamma_map test. All other tests should be passing except `mne/inverse_sparse/tests/test_mxne_inverse.py:test_mxne_inverse_standard`:
```
>       assert stc_prox.vertices[1][0] in label.vertices
E       IndexError: index 0 is out of bounds for axis 0 with size 0
```
*Then* the `loose_method='sum'` should change to `loose_method='svd'`, but this will break the sphere test. (The orientation only matches to 0.89, not 0.99, and the dipole fit is not 1.7mm away but 30mm :( ).

I'm out of my depth working on the MxNE stuff, so hopefully you find some hacking time soon.

Todo:
- [x] add `limit_depth_chs='whiten'` to MxNE and MNE
- [x] fix `test_mxne_inverse_standard`
- [x] ~~set `loose_method='svd'` and fix the failing sphere model test~~
- [x] add bias tests for LCMV

Next PR:
- rework `compute_depth_prior` to operate on `forward`

### CODE REVIEW

**Comment by larsoner** [comment]
@agramfort okay with you for a backward compat break to refactor `compute_depth_prior` to take `forward` instead of `gain, gain_info, is_fixed_orient, patch_areas`? Given that this function wasn't in the doc and buried in `mne.forward.forward.compute_depth_prior` I think it's probably fine, esp. if we note it well in the doc. Or should we do a proper deprecation cycle? We could allow `gain` to be an instance of `forward`, and have the other options be `None` and removed in 0.19. Let me know which you prefer.

If you want this refactoring/deprecation to appear in this PR, let me know. Otherwise this can be merged and the `compute_depth_prior` cleanups can be the next PR.

File: mne/beamformer/tests/test_lcmv.py [hunk 1]
```diff
@@ -690,4 +690,57 @@ def test_lcmv_reg_proj(proj):
     assert filters['rank'] == want_rank
 
 
+@pytest.mark.parametrize('reg, weight_norm, lower, upper', [
```
Comment (larsoner) [inline mne/beamformer/tests/test_lcmv.py]: @britta-wstnr since I reworked the test fixtures a bit in this PR it was easy for me to add the same "localization bias test" for LCMV that we have for the minimum norm solvers. Basically each "time point" in the `evoked` instance is a column of the gain matrix. I form `data_cov` by taking the dot product of this with its transpose and run LCMV on it with various options.

LCMV localization using this method (and it might not be ideally set up!) is perfect (100% of the ~500 dipoles localized correctly, according to the peak of the LCMV localization of each "time point" in the `evoked` instance) in the fixed case, and between 30 and 35% for most of the `free`-orientation inverses (similar to `MNE` performance in the `test_inverse.py` file).

Are you okay with these tests? It's entirely possible that they con be improved to get the percentages higher, but in the meantime they seem like a nice non-regression test since they ensure localization does not change (for better or worse) with changes, unless such changes are intended to change things.

WDYT?

File: mne/minimum_norm/tests/test_inverse.py [hunk 2]
```diff
@@ -299,97 +284,87 @@ def test_inverse_operator_channel_ordering():
...
+    # Compute the percentage of sources for which there is no localization
+    # bias:
+    perc = (want == np.argmax(loc, axis=0)).mean() * 100
+    assert lower <= perc <= upper, method
+
+
+@pytest.mark.parametrize('method, lower, upper, kwargs, depth', [
+    ('MNE', 35, 40, {}, dict(limit_depth_chs=False)),
+    ('MNE', 45, 55, {}, 0.8),
+    ('MNE', 65, 70, {}, dict(limit_depth_chs='whiten')),
```
Comment (larsoner) [inline mne/minimum_norm/tests/test_inverse.py]: FYI this shows the performance improvement for MNE when using `whiten`-based depth weighting (at least in this case, with this measure for localization bias)

File: mne/beamformer/tests/test_lcmv.py (hunk 1 above)
Comment (britta-wstnr) [reply mne/beamformer/tests/test_lcmv.py]: I am not sure I follow your explanation and the code. Isn't ``evoked`` an identity matrix? Can you point me to where the gain matrix magic happens for ``evoked``?
And do you mean ``data_cov`` is basically ``L * L'`` with ``L`` being the leadfield?
Sorry, I am a bit confused ...

Comment (larsoner) [reply mne/beamformer/tests/test_lcmv.py]: > Isn't `evoked` an identity matrix

Yes during construction and application of the `filters` (or for MNE, the `inv`) we have `evoked.data = np.eye(...)` to get us the output of the method to an identity input. We then take advantage of linearity (via an `einsum` or `dot` operation) to give the same output that would have been obtained as if we had just applied it to and Evoked instance that had `evoked.data = L`. It's just a mathematical trick to make it easier to compute the localization bias later.

> And do you mean `data_cov` is basically `L * L'` with `L` being the leadfield?

Yes exactly, this is what I meant to do but actually failed :) I've pushed a commit to fix it with updated percentages.

Comment (britta-wstnr) [reply mne/beamformer/tests/test_lcmv.py]: Mh, does this mean we leave data completely out of the pipeline?

Comment (britta-wstnr) [reply mne/beamformer/tests/test_lcmv.py]: (For _this_ test, anyway)

Comment (larsoner) [reply mne/beamformer/tests/test_lcmv.py]: Almost -- the only measured data used is the empirical `noise_cov` (unless `cov=False` in the param list, in which case it isn't even used). The other "data" is just the lead field, which depends just on sensor geometry (and the projections in the file, probably).

Comment (larsoner) [reply mne/beamformer/tests/test_lcmv.py]: Okay with you now?

Also I pushed a commit to simplify things -- no more `dot`, just actually have an EvokedArray with `fwd['sol']['data']` as the `data` directly to be more explicit.

Comment (britta-wstnr) [reply mne/beamformer/tests/test_lcmv.py]: Mh, for me it depends on the purpose of this test, I guess. Leaving the data out and using ``L * L'`` instead makes the "beamformer" very MNE-like. I guess that is okay for the purpose of detecting alarming changes in the code, but probably not a "proof-of-concept" for the beamformer - at least not fully.

File: mne/beamformer/tests/test_lcmv.py [hunk 3]
```diff
@@ -690,4 +690,67 @@ def test_lcmv_reg_proj(proj):
...
+@pytest.mark.parametrize('reg, weight_norm, use_cov, lower, upper', [
+    (0.05, 'unit-noise-gain', True, 100, 100),
+    (0.00, 'unit-noise-gain', True, 45, 55),
+    (0.05, 'nai', True, 100, 100),
+    (0.05, None, True, 100, 100),
+    (0.05, 'unit-noise-gain', False, 100, 100),
+])
+def test_localization_bias_fixed(bias_params, reg, weight_norm, use_cov,
+                                 lower, upper):
+    """Test inverse localization bias for loose minimum-norm solvers."""
```
Comment (agramfort) [inline mne/beamformer/tests/test_lcmv.py]: for LCMV you mean

File: conftest.py [hunk 4]
```diff
@@ -15,6 +16,20 @@
 # associated VTK import, so this could be another manifestation of that.
 import readline  # noqa
 
+import numpy as np
+import mne
+from mne.datasets import testing
+
+test_path = testing.data_path(download=False)
+s_path = op.join(test_path, 'MEG', 'sample')
+fname_data = op.join(s_path, 'sample_audvis_trunc-ave.fif')
```
Comment (agramfort) [inline conftest.py]: fname_data -> fname_evoked

File: mne/forward/forward.py [hunk 5]
```diff
@@ -1078,23 +1080,50 @@ def compute_depth_prior(G, gain_info, is_fixed_ori, exp=0.8, limit=10.0,
...
+            by the noise covariance (e.g., projections, SNR).
+
+    combine_xyz : 'spectral' | 'L2'
         When a loose (or free) orientation is used, how the depth weighting
         for each triplet should be calculated.
-        If 'svd', use the largest singular value of the 3x3 matrix
-        formed by the dot product ``Gk.T @ Gk`` will be used.
-        If 'sum', then ``np.sum(Gk ** 2)`` will be used.
+        If 'spectral', use the squared spectral norm of Gk.
+        If 'L2', use the squared L2 norm of Gk.
```
Comment (agramfort) [inline mne/forward/forward.py]: how about we call it 'fro' as it's the frobenius norm of the block of 3 cols. Like used in linalg.norm?

File: mne/minimum_norm/tests/test_inverse.py [hunk 6]
```diff
@@ -204,33 +196,30 @@ def _compare_io(inv_op, out_file_ext='.fif'):
     _compare(inv_init, inv_op)
 
 
-@testing.requires_testing_data
```
Comment (agramfort) [inline mne/minimum_norm/tests/test_inverse.py]: how does the fixture work? now every test requires the testing data?

**Comment by agramfort** [comment]
>
> @agramfort <https://github.com/agramfort> okay with you for a backward
> compat break to refactor compute_depth_prior to take forward instead of gain,
> gain_info, is_fixed_orient, patch_areas? Given that this function wasn't
> in the doc and buried in mne.forward.forward.compute_depth_prior I think
> it's probably fine, esp. if we note it well in the doc. Or should we do a
> proper deprecation cycle? We could allow gain to be an instance of forward,
> and have the other options be None and removed in 0.19. Let me know which
> you prefer.
>
> If you want this refactoring/deprecation to appear in this PR, let me
> know. Otherwise this can be merged and the compute_depth_prior cleanups
> can be the next PR.
>

if this is still relevant I let you judge. Maybe in a next PR if this is
pretty much ready.

File: mne/minimum_norm/tests/test_inverse.py (hunk 6 above)
Comment (larsoner) [reply mne/minimum_norm/tests/test_inverse.py]: Nope, ones that use the new `evoked` and `cov` fixtures (currently just these tests in `minimum_norm` and `lcmv`) are automatically marked as requiring the test data by the fixtures themselves.
